File: src/main/java/scrabble/phrases/providers/ComparisonProvider.java
package scrabble.phrases.providers;

import scrabble.phrases.dictionary.WordDictionary;
import scrabble.phrases.words.Adjective;
import scrabble.phrases.words.Noun;
import scrabble.phrases.words.NounGender;

/**
 * Provides comparison sentences.
 * Format: [Noun] e mai [adjective] decât [noun].
 */
public class ComparisonProvider extends SentenceProvider {

    public ComparisonProvider(WordDictionary dictionary) {
        super(dictionary);
    }

    @Override
    protected void initFilters() {
        // No filters needed
    }

    @Override
    public String getSentence() {
        Noun noun1 = getDictionary().getRandomNoun();
        Adjective adj = getDictionary().getRandomAdjective();
        Noun noun2 = getDictionary().getRandomNoun();

        String adjForm = (noun1.gender() == NounGender.F) ? adj.feminine() : adj.word();
        return noun1.articulated() + " e mai " + adjForm + " decât " + noun2.articulated() + ".";
    }
}

File: src/main/kotlin/scrabble/phrases/providers/ComparisonProvider.kt
package scrabble.phrases.providers

import scrabble.phrases.repository.WordRepository

class ComparisonProvider(
    private val repo: WordRepository,
    private val minRarity: Int = 1,
    private val maxRarity: Int
) : ISentenceProvider {

    override fun getSentence(): String {
        val noun1 = repo.getRandomNoun(minRarity = minRarity, maxRarity = maxRarity) ?: return "Ceva e mai mare decât altceva."
        val adj = repo.getRandomAdjective(minRarity = minRarity, maxRarity = maxRarity) ?: return "Ceva e mai mare decât altceva."
        val noun2 = repo.getRandomNoun(minRarity = minRarity, maxRarity = maxRarity, exclude = setOf(noun1.word)) ?: return "Ceva e mai mare decât altceva."

        return "${noun1.articulated} e mai ${adj.forGender(noun1.gender)} decâe ${noun2.articulated}."
    }
}
